hugr-py/src/hugr/_cond_loop.py

```diff
@@ -0,0 +1,99 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+
+import hugr._ops as ops
+
+from ._dfg import _DfBase
+from ._hugr import Hugr, Node, ParentBuilder, ToNode, Wire
+from ._tys import Sum, TypeRow
+
+
+class Case(_DfBase[ops.Case]):
+    _parent_cond: Conditional | None = None
+
+    def set_outputs(self, *outputs: Wire) -> None:
+        super().set_outputs(*outputs)
+        if self._parent_cond is not None:
+            self._parent_cond._update_outputs(self._wire_types(outputs))
+
+
+@dataclass
+class _IfElse(Case):
+    def __init__(self, case: Case) -> None:
+        self.hugr = case.hugr
+        self.parent_node = case.parent_node
+        self.input_node = case.input_node
+        self.output_node = case.output_node
+        self._parent_cond = case._parent_cond
+
+    def _parent_conditional(self) -> Conditional:
+        assert self._parent_cond is not None, "If must have a parent conditional."
+        return self._parent_cond
+
+
+class If(_IfElse):
+    def add_else(self) -> Else:
+        return Else(self._parent_conditional().add_case(0))
+
+
+class Else(_IfElse):
+    def finish(self) -> Node:
+        return self._parent_conditional().parent_node
+
+
+@dataclass
+class Conditional(ParentBuilder[ops.Conditional]):
+    cases: dict[int, Node | None]
+
+    def __init__(self, sum_ty: Sum, other_inputs: TypeRow) -> None:
+        root_op = ops.Conditional(sum_ty, other_inputs)
+        hugr = Hugr(root_op)
+        self._init_impl(hugr, hugr.root, len(sum_ty.variant_rows))
+
+    def _init_impl(self: Conditional, hugr: Hugr, root: Node, n_cases: int) -> None:
+        self.hugr = hugr
+        self.parent_node = root
+        self.cases = {i: None for i in range(n_cases)}
+
+    @classmethod
+    def new_nested(
+        cls,
+        sum_ty: Sum,
+        other_inputs: TypeRow,
+        hugr: Hugr,
+        parent: ToNode | None = None,
+    ) -> Conditional:
+        new = cls.__new__(cls)
+        root = hugr.add_node(
+            ops.Conditional(sum_ty, other_inputs),
+            parent or hugr.root,
+        )
+        new._init_impl(hugr, root, len(sum_ty.variant_rows))
+        return new
+
+    def _update_outputs(self, outputs: TypeRow) -> None:
+        if self.parent_op._outputs is None:
+            self.parent_op._outputs = outputs
+        else:
+            assert outputs == self.parent_op._outputs, "Mismatched case outputs."
+
+    def add_case(self, case_id: int) -> Case:
+        assert case_id in self.cases, f"Case {case_id} out of possible range."
```

Comment:
I think some of these `assert`s should be `raise`s instead (maybe `ValueError`?), since they occur through use of the API beyond our control.